### PR TITLE
[dv] Case 'CompareCtrConsistency' split cases '*Done' & '*Waiting'

### PR DESCRIPTION
**Pull request by antmarzam**
The sequence in case 'CompareCtrConsistency' tried to inject errors in
`rom_ctrl_compare.addr_q`. However, it didn't handle injecting errors
when the FSM of `rom_ctrl_compare` was in `state_q=Waiting` correctly.

The case has been split between `CompareCtrConsistencyWaiting` and
`CompareCtrConsistencyDone` and each of them handle error injection in
states `Waiting` and `Done` respectively.

For `CompareCtrConsistencyWaiting` the delay before injecting an error
has also been reduced by a fifth to increase the chances of the FSM in
`Waiting` state.

### CODE REVIEW

**Review by rswarbrick — COMMENTED** [review]
I'm not completely sure about this. I think that the CompareCtrConsistency check was supposed to be waiting until a time when we know `addr_q` was zero (8 modulo 8, I think).

I think the test you're looking at isn't working because that's not true! But it would probably be worth figuring out why `addr_q` is not zero.

File: hw/ip/rom_ctrl/dv/env/seq_lib/rom_ctrl_corrupt_sig_fatal_chk_vseq.sv [hunk 1]
```diff
@@ -10,16 +10,18 @@ class rom_ctrl_corrupt_sig_fatal_chk_vseq extends rom_ctrl_base_vseq;
   localparam int unsigned RomSizeWords = rom_ctrl_reg_pkg::ROM_CTRL_ROM_SIZE >> 2;
   localparam int unsigned RomIndexWidth = prim_util_pkg::vbits(RomSizeWords);
 
-  typedef enum bit [2:0] {
+  typedef enum bit [4:0] {
```
Comment (rswarbrick) [inline hw/ip/rom_ctrl/dv/env/seq_lib/rom_ctrl_corrupt_sig_fatal_chk_vseq.sv:13]: Silly nit: I think the list will fit in `[3:0]`?

File: hw/ip/rom_ctrl/dv/env/seq_lib/rom_ctrl_corrupt_sig_fatal_chk_vseq.sv [hunk 2]
```diff
@@ -105,25 +107,43 @@ class rom_ctrl_corrupt_sig_fatal_chk_vseq extends rom_ctrl_base_vseq;
           chk_fsm_state();
         end
         // The hash comparison module has an internal count. If this glitches to a nonzero value
-        // before the comparison starts or to a value other than the last index after the
-        // comparison ends then a fatal alert is generated.
-        CompareCtrConsistency: begin
+        // before the comparison starts (state=Waiting) or to a value other than the last index
+        // after the comparison ends (state=Done) then a fatal alert is generated.
+        // Each of these cases are covered in cases "CompareCtrConsistencyWaiting" and
```
Comment (rswarbrick) [inline hw/ip/rom_ctrl/dv/env/seq_lib/rom_ctrl_corrupt_sig_fatal_chk_vseq.sv:112]: Grammar nit: I don't think we need both "Each of ..." and "... respectively" here. Maybe just drop the first two words?

File: hw/ip/rom_ctrl/dv/env/seq_lib/rom_ctrl_corrupt_sig_fatal_chk_vseq.sv [hunk 3]
```diff
@@ -105,25 +107,43 @@ class rom_ctrl_corrupt_sig_fatal_chk_vseq extends rom_ctrl_base_vseq;
...
-          check_for_alert();
-          chk_fsm_state();
-          dut_init();
+          bit [4:0] fsm_state_q;
+
+          wait_with_bound(2000);
+          uvm_hdl_read("tb.dut.gen_fsm_scramble_enabled.u_checker_fsm.u_compare.state_q",
+                       fsm_state_q);
+          if(fsm_state_q == 5'b00100) begin
+            //State == Waiting -> It's ok to poke the addr_q to be non-zero
```
Comment (rswarbrick) [inline hw/ip/rom_ctrl/dv/env/seq_lib/rom_ctrl_corrupt_sig_fatal_chk_vseq.sv:122]: Nit: I think we normally stick a space after the `//`

(And similarly on line 139)

File: hw/ip/rom_ctrl/dv/env/seq_lib/rom_ctrl_corrupt_sig_fatal_chk_vseq.sv [hunk 4]
```diff
@@ -105,25 +107,43 @@ class rom_ctrl_corrupt_sig_fatal_chk_vseq extends rom_ctrl_base_vseq;
...
-          wait_with_bound(10000);
-          `DV_CHECK_STD_RANDOMIZE_WITH_FATAL(invalid_addr, invalid_addr > 0;);
-          force_sig("tb.dut.gen_fsm_scramble_enabled.u_checker_fsm.u_compare.addr_q",
-                    invalid_addr);
-          check_for_alert();
-          chk_fsm_state();
-          dut_init();
+          bit [4:0] fsm_state_q;
+
+          wait_with_bound(2000);
```
Comment (rswarbrick) [inline hw/ip/rom_ctrl/dv/env/seq_lib/rom_ctrl_corrupt_sig_fatal_chk_vseq.sv:118]: Why do we need a wait at all here? (I can believe that one might be necessary!)

If it's something uniform across the different cases, I wonder whether it would make sense to pull the "Wait up to 10 cycles" logic out of the case statement (and add a comment explaining why the wait is needed at all).

Not needed for this PR, but it might be a simple way to make the file easier to understand.

File: hw/ip/rom_ctrl/dv/env/seq_lib/rom_ctrl_corrupt_sig_fatal_chk_vseq.sv [hunk 5]
```diff
@@ -105,25 +107,43 @@ class rom_ctrl_corrupt_sig_fatal_chk_vseq extends rom_ctrl_base_vseq;
...
+          uvm_hdl_read("tb.dut.gen_fsm_scramble_enabled.u_checker_fsm.u_compare.state_q",
+                       fsm_state_q);
+          if(fsm_state_q == 5'b00100) begin
```
Comment (rswarbrick) [inline hw/ip/rom_ctrl/dv/env/seq_lib/rom_ctrl_corrupt_sig_fatal_chk_vseq.sv:121]: I'd suggest moving `uvm_fatal` up to here ("if fsm_state_q != 00100, cause a fatal error"). Then the main code doesn't need to sit in an `if` block any more.

File: hw/ip/rom_ctrl/dv/env/seq_lib/rom_ctrl_corrupt_sig_fatal_chk_vseq.sv [hunk 6]
```diff
@@ -105,25 +107,43 @@ class rom_ctrl_corrupt_sig_fatal_chk_vseq extends rom_ctrl_base_vseq;
...
-          force_sig("tb.dut.gen_fsm_scramble_enabled.u_checker_fsm.u_compare.addr_q",
-                    invalid_addr);
-          check_for_alert();
-          chk_fsm_state();
-          dut_init();
+          bit [4:0] fsm_state_q;
+
+          wait_with_bound(2000);
+          uvm_hdl_read("tb.dut.gen_fsm_scramble_enabled.u_checker_fsm.u_compare.state_q",
+                       fsm_state_q);
```
Comment (rswarbrick) [inline hw/ip/rom_ctrl/dv/env/seq_lib/rom_ctrl_corrupt_sig_fatal_chk_vseq.sv:120]: Hmm, this file doesn't check the return value from its calls to `uvm_hdl_read`. If you grep for the task name in the repo, you'll find that we usually look at the result (using a variety of styles!)

Would you mind doing a follow-up to add these checks? I seem to remember that there's a linting tool that complains without them.

File: hw/ip/rom_ctrl/dv/env/seq_lib/rom_ctrl_corrupt_sig_fatal_chk_vseq.sv [hunk 7]
```diff
@@ -257,6 +277,7 @@ class rom_ctrl_corrupt_sig_fatal_chk_vseq extends rom_ctrl_base_vseq;
   task force_sig(string path, int value);
     chk_hdl_path(path);
     uvm_hdl_force(path, value);
+    `uvm_info(`gfn, $sformatf("Setting path: %s to value=0x%0x",path,value), UVM_DEBUG)
```
Comment (rswarbrick) [inline hw/ip/rom_ctrl/dv/env/seq_lib/rom_ctrl_corrupt_sig_fatal_chk_vseq.sv:280]: I think I'd probably drop this change?

File: hw/ip/rom_ctrl/dv/env/seq_lib/rom_ctrl_corrupt_sig_fatal_chk_vseq.sv [hunk 8]
```diff
@@ -257,6 +277,7 @@ class rom_ctrl_corrupt_sig_fatal_chk_vseq extends rom_ctrl_base_vseq;
   task force_sig(string path, int value);
     chk_hdl_path(path);
     uvm_hdl_force(path, value);
```
Comment (rswarbrick) [inline hw/ip/rom_ctrl/dv/env/seq_lib/rom_ctrl_corrupt_sig_fatal_chk_vseq.sv:279]: This can go in a separate commit (and PR!), but I'd suggest dropping the `chk_hdl_path` entirely and instead checking the return value of `uvm_hdl_force`. I think it would probably be a bit clearer.

File: hw/ip/rom_ctrl/dv/env/seq_lib/rom_ctrl_corrupt_sig_fatal_chk_vseq.sv (hunk 1 above)
Comment (antmarzam) [reply hw/ip/rom_ctrl/dv/env/seq_lib/rom_ctrl_corrupt_sig_fatal_chk_vseq.sv:13]: Well spotted! Fixed

File: hw/ip/rom_ctrl/dv/env/seq_lib/rom_ctrl_corrupt_sig_fatal_chk_vseq.sv (hunk 7 above)
Comment (antmarzam) [reply hw/ip/rom_ctrl/dv/env/seq_lib/rom_ctrl_corrupt_sig_fatal_chk_vseq.sv:280]: I just added this statement since I struggled to pin-point the specific moment in which a value was hdl-forced.
I was running simulations with verbosity=UVM_DEBUG and still didn't see `uvm_hdl_force` marks on the log-file. 

I can leave it out, but if the sequence has any issues with driving the wrong value it'll be less clear which specific HDL path is the culprit. 

Happy to take it out if you still think that's best. Thoughts?

File: hw/ip/rom_ctrl/dv/env/seq_lib/rom_ctrl_corrupt_sig_fatal_chk_vseq.sv (hunk 4 above)
Comment (antmarzam) [reply hw/ip/rom_ctrl/dv/env/seq_lib/rom_ctrl_corrupt_sig_fatal_chk_vseq.sv:118]: I don't think it's required , but maybe helpful?
By leaving the `wait_with_bounds` we'll ensure we're not injecting the `addr_q` error at the same time.
The sequence originally used the `wait_with_bounds` function in multiple places. Including

File: hw/ip/rom_ctrl/dv/env/seq_lib/rom_ctrl_corrupt_sig_fatal_chk_vseq.sv [hunk 9]
```diff
@@ -105,25 +107,44 @@ class rom_ctrl_corrupt_sig_fatal_chk_vseq extends rom_ctrl_base_vseq;
...
+          bit [4:0] fsm_state_q;
+
+          wait_with_bound(2000);
+          uvm_hdl_read("tb.dut.gen_fsm_scramble_enabled.u_checker_fsm.u_compare.state_q",
+                       fsm_state_q);
+          if(fsm_state_q != 5'b00100) begin
+            `uvm_fatal(`gfn, {"Case:'CompareCtrConsistencyWaiting' hit when 'rom_ctrl_compare' ",
+                              "state!=Done hence sequence-case exits"})
+          end
+          else begin
```
Comment (rswarbrick) [inline hw/ip/rom_ctrl/dv/env/seq_lib/rom_ctrl_corrupt_sig_fatal_chk_vseq.sv:125]: Since `uvm_fatal` is fatal, you don't actually need the `else` here, I don't think?

File: hw/ip/rom_ctrl/dv/env/seq_lib/rom_ctrl_corrupt_sig_fatal_chk_vseq.sv [hunk 10]
```diff
@@ -105,25 +107,44 @@ class rom_ctrl_corrupt_sig_fatal_chk_vseq extends rom_ctrl_base_vseq;
...
+            // State == Waiting -> It's ok to poke the addr_q to be non-zero
+            `DV_CHECK_STD_RANDOMIZE_WITH_FATAL(invalid_addr, invalid_addr > 0; );
```
Comment (rswarbrick) [inline hw/ip/rom_ctrl/dv/env/seq_lib/rom_ctrl_corrupt_sig_fatal_chk_vseq.sv:127]: Nit: Spare space after the ;

File: hw/ip/rom_ctrl/dv/env/seq_lib/rom_ctrl_corrupt_sig_fatal_chk_vseq.sv [hunk 11]
```diff
@@ -105,25 +107,44 @@ class rom_ctrl_corrupt_sig_fatal_chk_vseq extends rom_ctrl_base_vseq;
...
+        CompareCtrConsistencyDone: begin
+          bit [2:0] invalid_addr;
+
           wait (cfg.rom_ctrl_vif.pwrmgr_data.done == MuBi4True);
+          // After cfg.rom_ctrl_vif.pwrmgr_data.done = True we're in Done state
           wait_with_bound(10);
           // LastAddr has been set to 7
-          `DV_CHECK_STD_RANDOMIZE_WITH_FATAL(invalid_addr,
-                                             (invalid_addr >= 0 && invalid_addr < 7););
+          `DV_CHECK_STD_RANDOMIZE_WITH_FATAL(invalid_addr, (invalid_addr < 7););
```
Comment (rswarbrick) [inline hw/ip/rom_ctrl/dv/env/seq_lib/rom_ctrl_corrupt_sig_fatal_chk_vseq.sv:143]: Nit: I don't think we need the parens around the `<` comparison. (Not wrong, but we may as well get rid of them!)

File: hw/ip/rom_ctrl/dv/env/seq_lib/rom_ctrl_corrupt_sig_fatal_chk_vseq.sv [hunk 12]
```diff
@@ -98,11 +99,12 @@ class rom_ctrl_corrupt_sig_fatal_chk_vseq extends rom_ctrl_base_vseq;
         CompareCtrlFlowConsistency: begin
           bit rdata;
           pick_err_inj_point();
+          path = "tb.dut.gen_fsm_scramble_enabled.u_checker_fsm.start_checker_q";
           do begin
-            uvm_hdl_read("tb.dut.gen_fsm_scramble_enabled.u_checker_fsm.start_checker_q", rdata);
+            `DV_CHECK(uvm_hdl_read(path, rdata));
```
Comment (rswarbrick) [inline hw/ip/rom_ctrl/dv/env/seq_lib/rom_ctrl_corrupt_sig_fatal_chk_vseq.sv:104]: I like the idea of using a variable (path) to avoid having to repeat the string!

I wonder whether it would be clearer to declare `path` inside the case statements instead, though. I think the total length ends up the same (`string path = "....";`) and there's no risk of a reader not knowing whether `path` has a sensible value at a given point.

File: hw/ip/rom_ctrl/dv/env/seq_lib/rom_ctrl_corrupt_sig_fatal_chk_vseq.sv [hunk 13]
```diff
@@ -273,19 +275,20 @@ class rom_ctrl_corrupt_sig_fatal_chk_vseq extends rom_ctrl_base_vseq;
...
     @(negedge cfg.clk_rst_vif.clk);
-    uvm_hdl_release(path);
+    `DV_CHECK(uvm_hdl_release(path));
   endtask: force_sig
 
   task chk_fsm_state();
     bit rdata_alert;
     bit [$bits(rom_ctrl_pkg::fsm_state_e)-1:0] rdata_state;
-    uvm_hdl_read("tb.dut.gen_fsm_scramble_enabled.u_checker_fsm.alert_o", rdata_alert);
+    string  path = "tb.dut.gen_fsm_scramble_enabled.u_checker_fsm.alert_o";
```
Comment (rswarbrick) [inline hw/ip/rom_ctrl/dv/env/seq_lib/rom_ctrl_corrupt_sig_fatal_chk_vseq.sv:287]: I don't think we're actually using the string more than once, so it's probably clearer just to pass the literal here. That also avoids having to set `path` to a different value in a few lines' time (which is slightly harder for a reader to understand).

File: hw/ip/rom_ctrl/dv/env/seq_lib/rom_ctrl_corrupt_sig_fatal_chk_vseq.sv [hunk 14]
```diff
@@ -294,9 +297,10 @@ class rom_ctrl_corrupt_sig_fatal_chk_vseq extends rom_ctrl_base_vseq;
   // Note that this assumes no loops and linear progression through the state enum entries.
   task wait_for_fsm_state_inside(ref rom_ctrl_pkg::fsm_state_e states_to_visit[$]);
     bit [$bits(rom_ctrl_pkg::fsm_state_e)-1:0] rdata_state;
+    string  path = "tb.dut.gen_fsm_scramble_enabled.u_checker_fsm.state_q";
```
Comment (rswarbrick) [inline hw/ip/rom_ctrl/dv/env/seq_lib/rom_ctrl_corrupt_sig_fatal_chk_vseq.sv:300]: Again, I'd probably be inclined to keep the literal here. But I suspect the line gets a bit long with the `DV_CHECK` too! If so, no worries! But there's an extra space after `string` which is probably worth dropping.

**Review by rswarbrick — COMMENTED** [review]
A couple of nitty comments, but this looks good!

File: hw/ip/rom_ctrl/dv/env/seq_lib/rom_ctrl_corrupt_sig_fatal_chk_vseq.sv (hunk 13 above)
Comment (antmarzam) [reply hw/ip/rom_ctrl/dv/env/seq_lib/rom_ctrl_corrupt_sig_fatal_chk_vseq.sv:287]: I was trying to avoid the line-length that's why I used the string variable instead of the literal value

File: hw/ip/rom_ctrl/dv/env/seq_lib/rom_ctrl_corrupt_sig_fatal_chk_vseq.sv (hunk 14 above)
Comment (antmarzam) [reply hw/ip/rom_ctrl/dv/env/seq_lib/rom_ctrl_corrupt_sig_fatal_chk_vseq.sv:300]: Same as above, the line ended up longer than 100 characters if I didn't use the variable

**Review by rswarbrick — APPROVED** [review]
Thanks!

**Comment by rswarbrick** [comment]
Sorry to be annoying, but I think this is still going to need a bit of squashing. And I've been annoying and have just caused you to need another rebase (by merging e2fb40c15f5fbcd2eb809c48cbab270feef1233b). Sorry!

Anyway, I'm hoping that you can use interactive rebase to expand the "sorting conflicts out" commit over the other two, so we should end up with something nice and clean.
